public/scripts/extensions/tts/alltalk.js

```diff
@@ -0,0 +1,938 @@
+
+    //##################//
+    // Preview AT Voice //
+    //##################//
+
+    async previewTtsVoice(voiceName) {
+        try {
+            // Prepare data for POST request
+            const postData = new URLSearchParams();
+            postData.append("voice", `${voiceName}.wav`);
+            // Making the POST request
+            const response = await fetch(`${this.settings.provider_endpoint}/api/previewvoice/`, {
+                method: "POST",
+                headers: {
+                    'Content-Type': 'application/x-www-form-urlencoded'
+                },
+                body: postData,
+            });
+            if (!response.ok) {
+                const errorText = await response.text();
+                console.error(`[previewTtsVoice] Error Response Text:`, errorText);
+                throw new Error(`HTTP ${response.status}: ${errorText}`);
+            }
+            // Assuming the server returns a URL to the .wav file
+            const data = await response.json();
+            if (data.output_file_url) {
+                // Use an audio element to play the .wav file
+                const audioElement = new Audio(data.output_file_url);
+                audioElement.play().catch(e => console.error("Error playing audio:", e));
+            } else {
+                console.warn("[previewTtsVoice] No output file URL received in the response");
+                throw new Error("No output file URL received in the response");
+            }
+
+        } catch (error) {
+            console.error("[previewTtsVoice] Exception caught during preview generation:", error);
+            throw error;
+        }
+    }
+
+    //#####################//
+    //  Populate ST voices //
+    //#####################//
+
+    async getVoice(voiceName, generatePreview = false) {
+        // Ensure this.voices is populated
+        if (this.voices.length === 0) {
+            // Fetch voice objects logic
+        }
+        // Find the object where the name matches voiceName
+        const match = this.voices.find(voice => voice.name === voiceName);
+        if (!match) {
+            // Error handling
+        }
+        // Generate preview URL only if requested
+        if (!match.preview_url && generatePreview) {
+            // Generate preview logic
+        }
+        return match; // Return the found voice object
+    }
+
+    //##########################################//
+    //  Generate TTS Streaming or call Standard //
+    //##########################################//
+
+    async generateTts(inputText, voiceId) {
+        try {
+            if (this.settings.at_generation_method === 'streaming_enabled') {
+                // For streaming method
+                const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;

```

Comment:
@Cohee1207 Ive tried this bit of code:

```
            if (this.settings.at_generation_method === 'streaming_enabled') {
                // For streaming method
                const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;
                console.log("Streaming URL:", streamingUrl);
    
                // Fetch the audio stream and return the response object
                const response = await fetch(streamingUrl);
                if (!response.ok) {
                    throw new Error(`HTTP ${response.status}: Failed to fetch audio stream`);
                }
                return response; 
            } else {
```

ST does play back, but not until generation is completed. In other words, its not actually streaming the audio, its waiting for a complete wav file to appear (as far as I can tell). Does this sound like expected behaviour from ST? Or have you any other suggestion on this? As Streaming would be a bit pointless otherwise. Im happy for any suggestions/help you can offer with this.

Reply:
It wont handle a response as a url e.g. Ive tried this way:

```
    if (this.settings.at_generation_method === 'streaming_enabled') {
        // Construct the streaming URL
        const streamingUrl = `${this.settings.provider_endpoint}/api/tts-generate-streaming?text=${encodeURIComponent(inputText)}&voice=${encodeURIComponent(voiceId)}.wav&language=${encodeURIComponent(this.settings.language)}&output_file=stream_output.wav`;
        console.log("Streaming URL:", streamingUrl);

        // Return the streaming URL directly
        return streamingUrl; 
```

and that results in:

```
Streaming URL: http://localhost:7851/api/tts-generate-streaming?text=Hmm...%20So%2C%20you%20believe%20someone%20close%20to%20you%20may%20have%20committed%20murder%3F%20And%20you%20wish%20for%20me%20to%20investigate%20the%20matter%20without%20their%20knowledge%3F&voice=Alan_Rickman.wav&language=en&output_file=stream_output.wav
index.js:455 Current TTS job for Sherlock Holmes completed.
index.js:422 Uncaught (in promise) TypeError: response.blob is not a function
    at addAudioJob (index.js:422:38)
    at tts (index.js:474:5)
```

Though I can alter index.js so that it can differentiate e.g. 

```
async function playAudioData(audioData) {
    if (currentAudioJob == null) {
        console.log('Cancelled TTS playback because currentAudioJob was null');
        return;
    }

    if (audioData instanceof Blob) {
        // Handle Blob object (existing logic)
        const reader = new FileReader();
        reader.onload = function (e) {
            const srcUrl = e.target.result;
            audioElement.src = srcUrl;
            audioElement.addEventListener('ended', completeCurrentAudioJob);
            audioElement.addEventListener('canplay', () => {
                console.debug('Starting TTS playback');
                audioElement.play();
            });
        };
        reader.readAsDataURL(audioData);
    } else if (typeof audioData === 'string') {
        // Handle string as a URL for streaming audio
        audioElement.src = audioData;
        audioElement.addEventListener('ended', completeCurrentAudioJob);
        audioElement.addEventListener('canplay', () => {
            console.debug('Starting TTS playback');
            audioElement.play();
        });
    } else {
        console.error(`playAudioData received invalid data type ${typeof audioData}`);
    }
}
```

and

```
async function addAudioJob(responseOrUrl) {
    if (typeof responseOrUrl === 'string') {
        // If the input is a string, assume it is a URL for streaming audio
        audioJobQueue.push(responseOrUrl);
        console.debug('Pushed streaming audio URL to queue.');
    } else if (responseOrUrl instanceof Response) {
        // Handle the response object as before
        const audioData = await responseOrUrl.blob();
        if (!audioData.type.startsWith('audio/')) {
            throw `TTS received HTTP response with invalid data format. Expecting audio/*, got ${audioData.type}`;
        }
        audioJobQueue.push(audioData);
        console.debug('Pushed audio job to queue.');
    } else {
        // Handle unexpected input types
        throw `TTS received invalid audio data type ${typeof responseOrUrl}`;
    }
}
```
Would you be ok with that change?

Reply:
I think I know exactly what happened here......I must have taken a copy of index.js literally the day or so before you made that change... Obviously Ive made other changes to index.js... so Ive been re-using that one, even though Ive updated ST to latest a few times. So thats why its not been working. I just didnt know you had made that change and Ive been re-using the old index.js. Problem solved it seems.

Ok, Ive added my bits again to index.js (the new version) Ill run a few sanity tests to make sure Im not missing anything,, and then probably Ill re-sync my fork and post a new PR over.